usbdev: Introducing a generic USB device driver API

bergzand opened the pull request:
### Contribution description

This PR adds usbdev, a common API for interfacing USB device peripherals, similar to netdev. For now this adds only the headers. An implementation for the Atmel sam devices is ready and is going to be PR't after a bit of cleanup as an example implementation.

I've tried to keep the API easy to use and to be able to fully use the hardware capabilities of the peripherals such as built in DMA.

See also the doxygen documentation in `drivers/include/usb/usbdev.h`

### Testing procedure

Only headers here, not much to test yet really.

### Issues/PRs references

![image](https://bergzand.net/misc/usbus.dot.svg)

aabadie commented:
Awesome !

> I don't know who is also interested in USB

I am so I added myself in the reviewer list. I can work later on the STM32 implementation part (@astralien3000 did some preliminary work for F4)

bergzand commented:
> I am so I added myself in the reviewer list. I can work later on the STM32 implementation part (@astralien3000 did some preliminary work for F4)

That would be awesome. I noticed that at least the stm32f446(re) has two USB interfaces, a full speed one and a high speed one with a few more features. I think that at the moment an implementation that is shared over the most devices among the F4 devices would be very nice to have.

Also having a second implementation with a different device class would be very useful to verify that the API is implementable with different devices.

bergzand commented:
@aabadie I've looked a bit at the reference manuals from different supported CPU's. From what I've found, the NRF52840, the EFM32 and the atmel sam devices all support some form of DMA'ing the packets from and to memory. 
The ST devices however seem to have a dedicated RAM for packet memory which might be troublesome to use with the current API model.

kaspar030 commented:
Awesome!

(BTW, are you aware of #3890?)

bergzand commented on drivers/include/usb/usbdev.h:
```diff
@@ -0,0 +1,312 @@
+    void *context;                          /**< Ptr to the thread context  */
+};
+
+typedef struct usbdev_driver {
```
Except that I'm using it [here], if I remove the name here I have to add it as a forward declaration somewhere at the top of this header.

[here]: https://github.com/RIOT-OS/RIOT/pull/9830/files/2ed14be21d5220549cb5f7c7f45a874391fbb2e0#diff-45003cfb94155ff1ae9366cf74649052R138

kaspar030 replied:
ok

bergzand commented:
> The problem with the first solution is that it highly relies on what the low-level hardware provides. And as you noticed with ST it cannot be generalized to all hardware on the market. So maybe the second solution could be a good trade-off.

Yes, I'm looking at the USB peripheral for the stm32f103 at the moment to check if moving the buffers to the peripheral solves the issue.

> What about adding read and write function pointers to the endpoint struct (instead of playing the address option) ? Maybe it would make the API more handy.

How would this work? What kind of arguments would you use here?

> I have another question regarding the expected driver API. I guess you plan to provide it via a periph implementation or something ?

At the moment I want to keep it similar to the way netdev/GNRC is set up. This way it is possible to have multiple different USB peripherals on a board. This is the case with the stm32f446re (One USB HS OTG and one USB FS OTG peripheral).

aabadie commented:
> At the moment I want to keep it similar to the way netdev/GNRC is set up.

There is a big difference with netdev: the USB feature is directly provided by an internal peripheral of the CPU, like gpio, i2c, spi, etc. Netdev drivers are built on top the `periph_<whatever>` interface.
That's why the usbdev interface should also be thought as an HAL on top of the CPU.

> This is the case with the stm32f446re (One USB HS OTG and one USB FS OTG peripheral).

Can both be used at the same time (on the same USB port) ? Looking at the nucleo-f446ze, only the USB OTG FS is supported. But that's true that the CPU can handle both.

bergzand commented:
> There is a big difference with netdev: the USB feature is directly provided by an internal peripheral of the CPU, like gpio, i2c, spi, etc. Netdev drivers are built on top the periph_<whatever> interface.
> That's why the usbdev interface should also be thought as an HAL on top of the CPU.

Most of the time yes, but [not necessarily always the case](https://www.maximintegrated.com/en/products/interface/controllers-expanders/MAX3420E.html). Anyway, Is this discussion about whether we want the header file in `drivers/include/periph`?

> Can both be used at the same time (on the same USB port) ? Looking at the nucleo-f446ze, only the USB OTG FS is supported. But that's true that the CPU can handle both.

From how I read it, the stm32f446ze offers two USB peripherals which both can be used at the same time. How the board implements this is of course a different matter.

dylad commented:
This looks promising !

> I guess you plan to provide it via a periph implementation or something ?

I think an USB device implementation should belong to periph_xxx rather than netdev.

Can't wait to try it on sam0 devices !

dylad commented:
Regarding VID/PID, maybe we can ask for a free PID [there](http://pid.codes/) 
Unless RIOT manages to get enough money to buy a VID (around 5k USD)...
The current VID here belongs to Logitech ;)

dylad commented:
In the current state, how can we associate an USBOPT to an endpoint ?
For example, if I want to generate a stall request to a specific endpoint. How can I get this endpoint in` int usbdev_set(usbdev_t *usbdev, usbopt_t opt, const void *value, size_t value_len)` function ?

bergzand commented:
> Regarding VID/PID, maybe we can ask for a free PID [there](http://pid.codes/)

I think this would be a great idea. Most of the time we just need a generic VID/PID combination because default drivers are able to handle everything and they don't listen to specific VID/PID combinations.

I think we would have to include a small bit of documentation refering the requirements of pid.codes specifying something like that the hardware must be open when it is used and (but more for our own protection) that the VID/PID combination must not be used as a basis for driver selection.

bergzand commented:
After working a bit with the stack, I've come to a few conclusions:

It doesn't seem to make sense to have a `usbdev_ep_event_cb_t` callback member for every `usbdev_ep_t` instance. A single callback in the `usbdev_t` struct should be enough as it can then be multiplexed out to different handlers based on the upper layer stack. The same can be applied to the `context` ptr, it can be removed from the `usbdev_ep_t` struct and it is probably sufficient to use the `usbdev_t` ptr directly.
The `usbdev_ep_driver_t` can probably also be moved to the `usbdev_t` struct. The case where a peripheral actually requires wildly different handling depending on the endpoint is the only reason to have a per endpoint struct ptr.

I'm thinking of merging the endpoint direction and number into a single `uint8_t` to match the usb protocol numbering, but I'd like to check if there is sufficient merit in merging these first.

bergzand commented:
Reworked a large part of the interface according to my own suggestion from above. I didn't rework the endpoint number/direction into a single `uint8_t`.

With this, there is only a single driver struct remaining covering both the usb device and the endpoint specific api. The `usbdev_ep_t` struct got stripped of the `context`, `driver` and `cb` members, those have been added/merged into the `usbdev_t` struct. I've added a `usbdev_t` ptr to the `usbdev_ep_t` struct. This simplifies the implementation quite a bit and prevents bugs where an `usbdev_ep_t` and `usbdev_t` from two different peripherals are used in a single call.

Helper functions for each call are also available and recommended to use. This:
 - prevents things such as `usbdev->dev->driver->init(usbdev->dev);` (actual example) and is used as `usbdev_init(usbdev->dev);` instead.
 - allows for a single point to add all preconditions for a function instead of each implementation having to add them.

bergzand commented:
rebased

bergzand commented:
Addressed CI complaints

kaspar030 commented:
This looks quite mature. I suggest squashing.

bergzand commented:
> This looks quite mature. I suggest squashing.

Squashed!

bergzand commented:
I've replaced the current default PID and VID combination with `#error` messages. This to prevent stalling this PR on a VID/PID discussion. While I agree to request a PID at least for development and testing purposes at some point, I'd rather not configure a [temporary pid] without at least a fat compile time warning for the end user.

So with this it is up to the developer to explicitly configure a VID and PID number. Having a default PID could be provided in a follow up

[temporary pid]: http://pid.codes/1209/0001/

bergzand commented:
#### Summarizing an out-of-band discussion with @kaspar030:
usbdev should be moved to periph_usbdev. The main advantage is to be able to reuse the periph feature handling. The preference is to keep the `usbdev_` names as not to clutter the `usb_` C function namespace unnecessarily. This also emphasizes that the peripheral is only a low level interface and not a fully functional USB stack.

@kaspar030 Did I forget something here?

kaspar030 commented:
> Did I forget something here?

no, perfect!

bergzand commented:
Moved `usbdev.h` to `drivers/include/periph` and adjusted the doxygen group. The `drivers_usbdev` group is also removed.

bergzand commented:
And updated the header guards accordingly

bergzand commented:
@kaspar030 @dylad any remaining blockers here? Can I squash a second time?

dylad commented:
You can squash. I'll review again on wednesday.

dylad commented on drivers/include/periph/usbdev.h:
```diff
@@ -0,0 +1,559 @@
+ * endpoints are allocated
+ */
+#ifndef USBDEV_NUM_ENDPOINTS
+#define USBDEV_NUM_ENDPOINTS       8
```
What about define it within each cpu/ ? This will force user to ensure the MCU can handle them. We might have surprise in the future if we keep a default value here.

bergzand replied:
The idea here was to have a global value to set the number of IN and OUT endpoints. This way, an application that only needs a limited number of endpoints can set the required number in a platform-independent way.

I agree that this might cause issues. We could move this to something like `cpu.h`, but I can't say I really like that solution. I don't mind keeping it this way at the moment and solve the issue in a follow up as soon as we have a few drivers.

bergzand commented:
That didn't exactly go as plan, but it should be okay now.

@dylad I've removed the `usbdev` pseudomodule, it is not necessary anymore now that it is a `periph` thing.

kaspar030 commented:
ACK from my side, too. &go!

bergzand commented:
Thanks guys!
